pkg/notebook/generate_test.go

```diff
@@ -0,0 +1,173 @@
+package notebook
+
+import (
+	"fmt"
+	"reflect"
+	"testing"
+
+	"github.com/tkestack/elastic-jupyter-operator/api/v1alpha1"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	JupyterNotebookName      = "jupyternotebook-sample"
+	JupyterNotebookNamespace = "default"
+	DefaultContainerName     = "notebook"
+	DefaultImage             = "busysandbox"
+	DefaultImageWithGateway  = "jupyter/base-notebook:python-3.8.6"
+	GatewayName              = "gateway"
+	GatewayNamespace         = "default"
+)
+
+var (
+	podSpec = v1.PodSpec{
+		Containers: []v1.Container{
+			{
+				Name:  DefaultContainerName,
+				Image: DefaultImage,
+			},
+		},
+	}
+
+	emptyNotebook = &v1alpha1.JupyterNotebook{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "kubeflow.tkestack.io/v1alpha1",
+			Kind:       "JupyterNotebook",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      JupyterNotebookName,
+			Namespace: JupyterNotebookNamespace,
+		},
+	}
+
+	notebookWithTemplate = &v1alpha1.JupyterNotebook{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "kubeflow.tkestack.io/v1alpha1",
+			Kind:       "JupyterNotebook",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      JupyterNotebookName,
+			Namespace: JupyterNotebookNamespace,
+		},
+		Spec: v1alpha1.JupyterNotebookSpec{
+			Template: &v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{"app": "notebook"},
+				},
+				Spec: podSpec,
+			},
+		},
+	}
+
+	notebookWithGateway = &v1alpha1.JupyterNotebook{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "kubeflow.tkestack.io/v1alpha1",
+			Kind:       "JupyterNotebook",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      JupyterNotebookName,
+			Namespace: JupyterNotebookNamespace,
+		},
+		Spec: v1alpha1.JupyterNotebookSpec{
+			Gateway: &v1.ObjectReference{
+				Kind: "JupyterGateway",
+			},
+		},
+	}
+
+	completeNotebook = &v1alpha1.JupyterNotebook{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "kubeflow.tkestack.io/v1alpha1",
+			Kind:       "JupyterNotebook",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      JupyterNotebookName,
+			Namespace: JupyterNotebookNamespace,
+		},
+		Spec: v1alpha1.JupyterNotebookSpec{
+			Gateway: &v1.ObjectReference{
+				Kind:      "JupyterGateway",
+				Namespace: GatewayNamespace,
+				Name:      GatewayName,
+			},
+			Template: &v1.PodTemplateSpec{
+				Spec: podSpec,
+			},
+		},
+	}
+)
+
+func TestGenerate(t *testing.T) {
+	// Test newGenerator
+	var err error
+	var g1, g2, g3, g4 *generator
+	_, err = newGenerator(nil)
```

Comment:
Maybe we can use table-driven style to write the tests. https://dave.cheney.net/2019/05/07/prefer-table-driven-tests